File: socketpy/cliente.py
import socket
import struct

def write_utf(sock, mensaje):# Convierte el mensaje a bytes UTF-8
    mensaje_utf8 = mensaje.encode('utf-8')
    
    
    longitud = len(mensaje_utf8)# Obtiene la longitud del mensaje en bytes
    
    
    encabezado = struct.pack('>H', longitud)  # convierte el número en 2 bytes para decirle al servidor cuánto mide el mensaje que voy a enviar
    # es un protocolo de comunicación que indica la longitud del mensaje que se va a enviar
    # '>H' significa que se espera un número de 2 bytes (sin signo) en orden grande primero (big-endian), pq estamos trabajando con java en e servidor
    
    sock.sendall(encabezado + mensaje_utf8)# Envía primero la longitud y luego el mensaje codificado
 
 # recibe los primeros 2 bytes, que indican la longitud del mensaje
def read_utf(sock):
   
   # lee (recibe) los primeros 2 bytes del mensaje enviado por el servidor.
   # esos 2 bytes indican la longitud del mensaje que viene después.
    encabezado = sock.recv(2)
    if not encabezado:
        return ''  # si no se recibe nada, retorna cadena vacía
    longitud = struct.unpack('>H', encabezado)[0]  # convierte los 2 bytes recibidos en un número. 
    # '>H' significa que se espera un número de 2 bytes (sin signo) en orden grande primero (big-endian). pq estamos trabajando con java en e servidor
    # [0] es porque devuelve una lista con un solo número, y tomamos ese número.

    datos = b''  # crea una variable vacía para ir guardando los datos recibidos en formato binario

    # recibe hasta completar el mensaje de la longitud indicada
    while len(datos) < longitud:
        parte = sock.recv(longitud - len(datos))
        if not parte:
            break
        datos += parte
    
    # decodifica los bytes recibidos a UTF-8 y retorna el mensaje como string
    return datos.decode('utf-8')

# Configuración del host y puerto al que se conecta el cliente
HOST = 'localhost'
PUERTO = 2566

# crea un socket con IPv4 (AF_INET) y protocolo TCP (SOCK_STREAM)
# el 'with' asegura que el socket se cierre automáticamente al salir del bloque
with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    s.connect((HOST, PUERTO))  # Conecta al servidor
    print("Conectado al servidor.")

    # el menu mientras esta en verdadero
    while True:
        print("\n*******MENÚ-python********")
        print("1. Generar nombre de usuario")
        print("2. Generar dirección de correo electrónico")
        print("3. Salir")
        opcion = input("Opción: ")

        if opcion == '1':
            # pide la longitud y manda el comando USERNAME con esa longitud
            longitud = input("Ingrese la longitud del nombre de usuario (5-20): ")
            write_utf(s, f"USERNAME:{longitud}")# f de formato string
            respuesta = read_utf(s)
            print("Servidor:", respuesta)

        elif opcion == '2':
            # pide el nombre de usuario y manda el comando EMAIL con ese nombre
            nombre = input("Ingrese su nombre de usuario: ")
            write_utf(s, f"EMAIL:{nombre}")# f de formato string
            respuesta = read_utf(s)
            print("Servidor:", respuesta)

        elif opcion == '3':
            # manda el comando SALIR para terminar la conexión
            write_utf(s, "SALIR")
            respuesta = read_utf(s)
            print("Servidor:", respuesta)
            print("Saliendo del cliente...")
            break

        else:
            print("Opción inválida.")

File: javaSockets/socket_cliente_servidor/src/socket_cliente_servidor/Cliente.java
package socket_cliente_servidor;

import java.io.*;
import java.net.*;//la clase socket viene incorporada desde esta libreria
import java.util.Scanner;

public class Cliente {

    public static void main(String[] args) {

        final String HOST = "localhost";//localhost indica donde se esta corriendo el programa
        final int PUERTO = 2566;

        try (
            Socket socket = new Socket(HOST, PUERTO);//se conecta con el server
        		DataOutputStream out = new DataOutputStream(socket.getOutputStream());
        		// Se crea un canal de salida llamado "out" a partir del socket. Esto significa que "out" es un flujo por donde el cliente puede enviar datos al servidor.El método socket.getOutputStream() devuelve el flujo básico para enviar bytes,
        		// y DataOutputStream lo envuelve para facilitar escribir datos con writeUTF.

            DataInputStream in = new DataInputStream(socket.getInputStream());
        		// Se crea un canal de entrada llamado "in" a partir del socket.Esto significa que "in" es un flujo por donde el cliente puede recibir datos que envía el servidor.
        		//El método socket.getInputStream() devuelve el flujo básico para leer bytes, y DataInputStream lo envuelve para facilitar la lectura de datos primitivos y cadenas (por ejemplo, con readUTF).

            Scanner teclado = new Scanner(System.in)//seinstancia para leer de la consola lo que se manda por teclado
        ) {
            System.out.println("Conectado al servidor.");//si esto funciono servidor conectado

            boolean continuar = true;//pone un true en el teclado para que ingrese al menu

            while (continuar) {
                System.out.println("\nOpciones: cliente java");
                System.out.println("1. Generar nombre de usuario");
                System.out.println("2. Generar correo electrónico");
                System.out.println("3. Salir");
                System.out.print("Elija una opción: ");
                int opcion = teclado.nextInt();//lee la opcion ingresada por teclado
                teclado.nextLine(); // limpiar buffer por el salto de linea

                switch (opcion) {
                    case 1:
                        System.out.print("Ingrese longitud (5 a 20): ");
                        int longitud = teclado.nextInt();//lee la longitud
                        teclado.nextLine(); // limpiar buffer por el salto de linea
                        out.writeUTF("USERNAME:" + longitud);// manda al servidor la cadena mas la longitud para que la valide con el whiteUTF 
                        System.out.println("Servidor: " + in.readUTF());//el servidor responde en forma utf
                        break;

                    case 2:
                        System.out.print("Ingrese nombre de usuario: ");
                        String nombre = teclado.nextLine();//lee el nombre del usuario
                        out.writeUTF("EMAIL:" + nombre);// manda al servidor la cadena emai mas el nombre para que la valide con el whiteUTF 
                        System.out.println("Servidor: " + in.readUTF());//el servidor responde en forma utf
                        break;

                    case 3:
                        out.writeUTF("SALIR");// manda al servidor la cadena  para que la valide con el whiteUTF  
                        System.out.println("Servidor: " + in.readUTF());//el servidor responde en forma utf
                        continuar = false;//corta el while
                        break;

                    default:
                        System.out.println("Opción no válida.");//por si hay una opcion no valida
                }
            }

        } catch (IOException e) {
            System.err.println("Error en el cliente: " + e.getMessage());//manda la excepcio a la consola o error
        }
    }
}
